hardware chapter (#84)
vblanco20-1/vulkan-guide: docs/extra-chapter/hardware.md

## Out of order execution

> All of this cache behavior brings a big amount of randomness into how long an instruction takes to load. The same `mov` instruction could take 10 cycles if it was on L1, but it could also take 200 if it was on RAM. During that time, the CPU would need to completely stop executing until that memory arrives. This happened on the PS3 cpu and before, and was one of its main sources of slowdown even while that cpu had a fast clock speed.
> To deal with this, CPUs began to execute instructions out of order. This way the pipeline of the cpu can continue doing useful work even if one of the instructions blocks for some time. Once out of order execution starts being a thing, the CPU can also improve performance by having multiple execution units for the different stages, and extract parallelism by itself. Modern CPUs can enqueue hundreds of instructions and schedule them for execution.
> 
> For example, one single CPU core having 3 separate ALUs. One ALU is running a long operation like a divide, while the other ALUs are doing quick 1-cycle adds. But when doing that we run into a problem, which is that some instructions end at different times, and some instructions also depend on each other. This is handled internally in the CPU through some queing systems, where instructions get read multiple at a time, and then after they are decoded they get into a pending queue until their dependencies are met. Due to the low amount of registers used, the registers would constantly conflict with each other, so hardware designers added a "register renamer" to the CPU that connects the low amount of registers in the code into a much bigger amount of registers in the cpu to improve the way the instructions can overlap. Once a instruction has met its dependencies, it gets added into the different execution units of the cpu, such as a ALU that does adds/sub only, or an ALU that does divides too, or one of multiple memory loaders. 
> 
> While instructions run at different amount of clock cycles and they also finish out of order, the cpu keeps track of an output list which will perform things like memory writes and atomic operations fully preserving the order of the code so that it works as written.
> 
> Some info can be seen on this wikipedia page [4]

Reviewer: ```suggestion
All of this cache behavior brings a big amount of randomness into how long an instruction takes to load. The same `mov` instruction could take 10 cycles if it was in L1, but could take 200 cycles if it has to access RAM. During that time, the CPU would need to completely stop executing that instruction until the memory arrives. This happened to the PS3 CPU, and was one of its main sources of slowdowns, even though that CPU had a fast clock speed.
To deal with this, CPUs began to execute instructions out of order. This way the pipeline of the CPU can continue doing useful work even if one of the instructions has to wait for some time. Once out of order execution starts being a thing, the CPU can also improve performance by having multiple execution units for the different stages, and extract parallelism by itself. Modern CPUs can enqueue hundreds of instructions and schedule them for execution.

For example, one single CPU core could have 3 separate ALUs. One ALU is running a long operation like a divide, while the other ALUs are doing quick 1-cycle adds. But when doing that we run into a problem, which is that some instructions end at different times, and some instructions also depend on each other. This is handled internally in the CPU through queuing systems, where multiple instructions are read at a time, then after they are decoded they get put into a pending queue until their dependencies are met. Due to instructions using a low amount of registers, the registers required by different queued instructions would constantly conflict with each other, so hardware designers added a "register renamer" to the CPU that connects the low amount of registers used in the code into a much bigger amount of registers that the CPU has to improve instructions overlap. Once a instruction has met its dependencies, it gets added into the different execution units of the CPU, such as a ALU that does adds/sub only, or an ALU that does divides too, or one of multiple memory loaders. 

While instructions run at different amount of clock cycles and they also finish out of order, the CPU keeps track of an output list which will perform things like memory writes and atomic operations fully preserving the order of the code so that it works as written.

Some info can be seen on this wikipedia page [4]
```

> ## Branch prediction
> With so much deep pipelining and out of order execution inside the CPU, we run into an issue with branches. What happens when the CPU hits a branch but the data is still not known on where to branch? We could stop the whole thing until the data arrives, but that would mean stopping the cpu completely for many clocks, if the branch data depends on a RAM load that might take 200 cycles to arrive. The modern CPUs decide to just decide on the branch direction, and continue executing anyway. If they decide wrong, they will flush the half-done instructions and the pending memory writes, and begin anew. If they guessed correctly all is good and nothing is done. This is known as branch prediction. It tries to predict the future to do educated guesses of where the program flow will go. To do that it keeps storage inside the CPU that records wether a given branch was taken or not last time, and even if there is a pattern like flipping between taken and not taken. The complexity of modern branch predictors is really high, and their specific working details are often closely guarded trade secrets. On something like a i7 or ryzen, the branch predictor hits the correct prediction more than 95% of the time on average. The place where they struggle is when the branch is essentially random for branches that are predictable and stable the predictor will almost always hit it correctly. If the branch is random and is constantly mispredicted, the CPU will stall and have to reset constantly due to the bad predictions, which is often a thing that can slow softare by a significant factor. In some cases, its worth to run both sides of the branch and use a select instruction or a branchless blend as the penalty for misprediction is higher than the calculation. For a video that talks about how to program in a branchless way for performance, "Branchless programming in Cpp" [5] can be a good one. Another great article that explains different types of branch predictor seen in CPUs is this one [14].

Reviewer: ```suggestion
## Branch prediction
With so much deep pipelining and out of order execution inside the CPU, we run into an issue with branches. A branch happens when the next instruction to execute is decided by the output data of a previous instruction. What happens when the CPU hits a branch but the data is still not known on where to branch? We could stop the whole thing until the data arrives, but that would mean stopping the CPU completely for many clocks. If the branch data depends on a RAM load, that might take 200 cycles to arrive! Modern CPUs instead decide to just pick one of the directions a branch could go, and continue executing anyway. If they decide wrong, they will flush the half-done instructions and any pending memory writes, and begin anew. If they guessed correctly all is good and nothing extra needs to be done. This is known as branch prediction. It tries to predict the future by doing educated guesses of where the program flow will go. To do that it keeps storage inside the CPU that records whether a given branch was taken or not last time, and even if there is a pattern like flipping between taken and not taken. The complexity of modern branch predictors is really high, and their specific working details are often closely guarded trade secrets. On something like a i7 or Ryzen, the branch predictor hits the correct prediction more than 95% of the time on average. The place where CPUs struggle is when the branch is essentially random. For branches that are predictable and stable, the predictor will almost always hit it correctly. If the branch is random and is constantly mispredicted, the CPU will stall and have to reset constantly due to the bad predictions, which is often a thing that can slow software by a significant factor. In some cases, its worth to run both sides of the branch and use a select instruction or a branchless blend as the penalty for misprediction is higher than the calculation. For a video that talks about how to program in a branchless way for performance, "Branchless programming in Cpp" [5] can be a good one. Another great article that explains different types of branch predictor seen in CPUs is this one [14].
```

## SIMD

> All of the above is focused on the core of executing instructions faster. But what if we want to do more work per instruction? Thats where SIMD comes in (Single Instruction, Multiple Data). Due to the needs of getting more work done per instruction, CPUs started adding new instructions that would do multiple operations at once. For example, normally a CPU will add numbers 1 at a time, but if you use AVX512 instruction set (on modern server i7s and last generation ryzens) they will execute 16 adds within one instruction, providing 16x the work per instruction. 
> SIMD generally has considerable limitations due to them being super complicated instructions (due to doing multiple things per instruction) and compilers often really struggle with them. Autovectorization has been bleeding edge research since SIMD instructions were first invented, and even today, trying to rely on autovectorization to maximize performance of a given piece of code is unrealiable. To get max performance, programmers often need to go into intrinsics (using the SIMD instructions directly in the code) to do their algorithms. There are also languages like ISPC[13] that let you write vectorized code.
> Over the years, there has been a progression on SIMD instruction sets. We first began with MMX and SSE instruction sets, which were 4-wide, so they ran 4 operations at a time. This was pretty useful for a lot of multimedia and graphics work as you deal with native vec4s very often. For example, the DXMath library (similar to GLM but for directX) uses SSE instructions a lot to accelerate vector and matrix math for game development, which uses vec4s and mat4x4s. As time continued, we then went to AVX instruction set, which worked 8 wide. Now the vectors dont fit the typical vec3s and vec4s used in graphics math, so using them is harder. Also AVX is incompatible with SSE, so the code needs to be rebuilt if you want to use the new instructions and wider execution. AVX 2 was released a few years later, which is still 8x, but it has a few better operations that are useful. 
> A few years ago, we got AVX512, which offers 16 wide execution, alongside a very advanced set of operations. Intel decided that gamers dont need it, so on their last generation i7 they removed it. At the time of writing this article, only server-class intel GPUs, and ryzen cpus support it, which means its a complete nonstarter to be used for game development as according to Steam, less than 30% of players have CPUs that support AVX512.
> On ARM platforms, you can find NEON support which is 4 wide but a nicer than the x86 PC SSE equivalent, and on the bleeding edge ARM cores for datacenters you can find SVE which is a vector instruction set that allows variable lenght vectors so that future cpus can have wider SIMD units and the code will still work. 
> As there are so many incompatible SIMD feature sets, creating intrinsics manually can be a significant developer effort because the developer would often need to write the code a minimum of 3 times. One for scalar normal code, other for AVX for PC and big consoles, and other for NEON for ARM (nintendo switch, phones). Or even multiple feature levels for PC alone.
> A good video on intrinsics is "The art of SIMD programming"[6] . If you dont want to use intrinsics and prefer a library, xSimd[12] can be a good one, where it uses templates to abstract multiple types of intrinsics

Reviewer: ```suggestion
All of the above is focused on the core of executing instructions faster. But what if we want to do more work per instruction? Thats where SIMD comes in (Single Instruction, Multiple Data). Due to the needs of getting more work done per instruction, CPUs started adding new instructions that would do multiple operations at once. For example, normally a CPU will add numbers 1 at a time, but if you use AVX512 instruction set (on modern server i7s and last generation Ryzens) they will execute 16 adds within one instruction, providing 16x the work per instruction. 
SIMD generally has considerable limitations due to them being super complicated instructions (due to doing multiple things per instruction) and compilers often really struggle with them. Autovectorization has been bleeding edge research since SIMD instructions were first invented, and even today, trying to rely on autovectorization to maximize performance of a given piece of code is unreliable. To get max performance, programmers often need to use intrinsics (using the SIMD instructions directly in the code) to do their algorithms. There are also languages like ISPC [13] that let you write vectorized code directly.
Over the years, there has been a progression on SIMD instruction sets. We first began with MMX and SSE instruction sets, which were 4-wide, so they ran 4 operations at a time. This was pretty useful for a lot of multimedia and graphics work as you deal with native vec4s very often. For example, the DXMath library (similar to GLM but for directX) uses SSE instructions a lot to accelerate vector and matrix math for game development, which uses vec4s and mat4x4s. As time continued, we then went to AVX instruction set, which worked 8 wide. Now the vectors don't fit the typical vec3s and vec4s used in graphics math, so using them is harder. Also AVX is incompatible with SSE, so the code needs to be rebuilt if you want to use the new instructions and wider execution. AVX 2 was released a few years later, which is still 8x, but it has a few better operations that are useful. 
A few years ago, we got AVX512, which offers 16 wide execution, alongside a very advanced set of operations. Intel decided that gamers don't need it, so on their last generation i7 they removed it. At the time of writing this article, only server-class intel CPUs, and Ryzen cpus support it, which means its a complete nonstarter to be used for game development as according to Steam, as less than 30% of players have CPUs that support AVX512.
On ARM platforms you can find NEON support, which is 4 wide but a bit nicer than the x86 PC SSE equivalent, and on the bleeding edge ARM cores for datacenters you can find SVE which is a vector instruction set that allows variable length vectors so that future CPUs can have even wider SIMD units and current code will still work. 
As there are so many incompatible SIMD feature sets, creating intrinsics manually can be a significant developer effort because the developer would often need to write the code a minimum of 3 times. One for scalar normal code, other for AVX for PC and big consoles, and other for NEON for ARM (Nintendo switch and phones). Or even multiple feature levels for PC alone.
A good video on intrinsics is "The art of SIMD programming"[6] . If you don't want to use intrinsics and prefer a library, xSimd[12] can be a good one, where it uses templates to abstract multiple types of intrinsics.
```

## Materials

> A lot of what is talked here is taken into account by the programming style known as data oriented programming. In that style of programming, developers generally write the code with things like the cache and branch predictor in mind, often placing their data into arrays and processing them in batches. The main book that talks about it is Data Oriented Design[7], and talks "Data oriented design and C++" [8] and "OOP is dead, long live data oriented design"[9] are classic ones about it. 
> 
> For information on the hardware details specifically, a great source of information is the Agner Fog manuals [10]. The first one is a must read, and the microarchitecture ones talk about a lot of the topics explained here. 
> Other source is Wikichip which can have good information on the details of different processors. The digital book "Algorithms for Modern Hardware" [15] is also great read that goes in depth on many of these concepts, with specific focus on programming around them and code examples for high performance algorithms.  

Reviewer: ```suggestion
A lot of what is talked here is taken into account by the programming style known as data oriented programming. In that style of programming, developers generally write the code with things like the cache and branch predictors in mind, often placing their data into arrays and processing them in batches. The main book that talks about it is Data Oriented Design[7], and the talks "Data oriented design and C++" [8] and "OOP is dead, long live data oriented design"[9] are classic ones about it. 

For information on specific hardware details, a great source of information is the Agner Fog manuals [10]. The first one is a must read, and the microarchitecture ones talk about a lot of the topics explained here. 
Other source is Wikichip which has good information on the details of different processors. The digital book "Algorithms for Modern Hardware" [15] is also great read that goes in depth on many of these concepts, with specific focus on programming around them and code examples for high performance algorithms.  
```
